stdenvAdapters.useLibsFrom: use targetStdenv.cc.override

ConnorBaker opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

`stdenvAdapters.useLibsFrom` now uses `targetStdenv.cc.override`, assuming that the `cc` attribute of the provided environment is a wrapped compiler. Beyond the benefits of avoiding a roundtrip wrapping, under cross-compilation we now use the copy of `coreutils` available to `targetStdenv`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

CC @NixOS/cuda-maintainers 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ConnorBaker commented:
> [!Note]
>
> Template `nixpkgs-review` command:
>
> ```console
> PR=281371; \
> SYSTEM="aarch64-linux"; \
> CUDA_SUPPORT="true"; \
> CUDA_CAPABILITIES='[ "7.5" ]'; \
> nixpkgs-review pr "$PR" \
>   --system "$SYSTEM" \
>   --no-shell \
>   --checkout commit \
>   --allow aliases \
>   --build-args "--max-jobs 1" \
>   --skip-package-regex 'python3(\d){1,2}Packages\.(pytorch-pfn-extras|ffcv)' \
>   --skip-package-regex 'python3(\d){1,2}Packages\.(shap|mlflow|optuna)' \
>   --extra-nixpkgs-config "{
>     allowUnfree = true;
>     allowBroken = false;
>     cudaSupport = ${CUDA_SUPPORT:-false};
>     cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
>   }"
> ```
>
> The packages `python3Packages.{pytorch-pfn-extras,shap}` are excluded because their checkPhase runs for multiple hours -- other excluded packages are dependent on these, and `nixpkgs-review` does not handle transitive dependencies, so we must find them ourselves.
>
> Template log archival command:
>
> ```bash
> PR=281371; \
> RUN_NUMBER=13; \
> SYSTEM="aarch64-linux"; \
> CUDA_CAPABILITIES="7_5"; \
> tar -cvf "$SYSTEM-cap-$CUDA_CAPABILITIES-pr-$PR-$RUN_NUMBER.tar.zst" \
>   -I "zstd -T0 --ultra -22" \
>   -C "$HOME/.cache/nixpkgs-review/pr-$PR-$RUN_NUMBER" \
>   .
> ```
>
> Logs are made available at <https://drive.google.com/drive/folders/1GgABhwa2ooKeZXMqf5He6PcyClNrE6cQ?usp=share_link>

## ✅ `aarch64-darwin`

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

## ✅ `x86_64-darwin`

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

## ✅ `aarch64-linux`

### ✅ CUDA (Jetson)

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = true;
    cudaCapabilities = [ "7.5" ];
  }'` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

### ✅ CUDA (Non-Jetson)

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = true;
    cudaCapabilities = [ "7.2" ];
  }'` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

### ✅ Non-CUDA

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

## ✅ `x86_64-linux`

### ✅ CUDA (Non-Jetson)

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = true;
    cudaCapabilities = [ "7.2" ];
  }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

### ✅ Non-CUDA

Result of `nixpkgs-review pr 281371 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

SomeoneSerge commented:
Interesting! I now wonder what would `buildPackages.wrapCCWith` do in the `useLibsFrom` definition

SomeoneSerge commented on pkgs/stdenv/adapters.nix:
Making a single attrset prevents one from using the partially applied form (`useGoodLibs = useLibsFrom stdenv; compatStdenv = useGoodLibs gcc10Stdenv`). The partially applied form with two attrsets might be a bit mouthful

SomeoneSerge replied:
What annoys me a bit is that the current form doesn't let one pass a cxxStdlib directly, in place of `modelStdenv`

`useLibsFrom = { stdenvForLibs ? null, cxxStdlib ? stdenvForLibs.cc.cc.lib }: targetStdenv: ...` would...

SomeoneSerge commented on pkgs/stdenv/adapters.nix:
I guess the comment we could add is something "we want to use coreutils from the same package set/slice as gcc"

Commit:
stdenvAdapters.useLibsFrom: use targetStdenv.cc.override

As @SomeoneSerge pointed out in https://github.com/NixOS/nixpkgs/pull/281371#discussion_r1531174155,
by avoiding `wrapCCWith` and using `targetStdenv.cc.override`, we avoid roundtrip wrapping and
are able to use `coreutils` from `targetStdenv`.

ConnorBaker commented:
Updated, rebased, and renamed the PR.

@SomeoneSerge can you give it another round of review? I minimized the changes so all this PR does is pass `coreutils` from `targetStdenv`.

SomeoneSerge commented on pkgs/stdenv/adapters.nix:
```diff
@@ -245,6 +246,11 @@ rec {
+        /* NOTE: We must provide coreutils explicitly so that wrapCCWith
+         * doesn't take it from pkgs, which is pkgsHostTarget, because it
+         * would fail to run on the buildPlatform when cross-compiling. */
+        coreutils = targetStdenv.cc.coreutils_bin;
```
Should we do this, or should we add another passthru? I'm uncomfortable with relying on `optionalString` not enforcing types

ConnorBaker replied:
I think it's fine, pending a larger rewrite/refactoring of the whole interface (as you've noted in #282356 and #283517. Is that agreeable to you? I'd like to get this merged.

SomeoneSerge commented on pkgs/stdenv/adapters.nix:
```diff
@@ -245,6 +246,11 @@ rec {
+        /* NOTE: We must provide coreutils explicitly so that wrapCCWith
+         * doesn't take it from pkgs, which is pkgsHostTarget, because it
+         * would fail to run on the buildPlatform when cross-compiling. */
+        coreutils = targetStdenv.cc.coreutils_bin;
```
It just occurred to me that instead of calling `wrapCCWith` directly we could assume that `targetStdenv.cc` is already a wrapper, and then we could take `targetStdenv.cc.override` with the same arguments

SomeoneSerge replied:
It's not a great assumption but it might be better than what we do now

SomeoneSerge replied:
Btw note https://github.com/NixOS/nixpkgs/issues/283517 if you haven't seen it yet

ConnorBaker replied:
Mhm, good observation about avoiding roundtrip wrapping. I'm curious if it'll break anything in Nixpkgs as I've implemented it -- it looks like `pythonPackages.tensorflow`, `gnuradio3_8`, and `cudaPackages.backendStdenv` are the only users within Nixpkgs.

I had seen the issue you opened -- I agree that there needs to be a better interface, but for the very short term work I have over the next two weeks, I think it's out of scope and I'd rather focus on this aspect of it. Does that make sense?

SomeoneSerge replied:
> I'm curious if it'll break anything in Nixpkgs as I've implemented it

Within Nixpkgs it should break strictly less (e.g. the previous implementation definitely could not be compatible with various `nixpkgsFun` instances like `pkgsStatic`, because it was just discarding the old wrapper's information). The exact interaction of `useLibsFrom` and custom `wrapCCWith` invocations, and the impact out of tree are still unclear.

> I agree that there needs to be a better interface, but for the very short term work I have over the next two weeks

I agree with switching to `.override` now, but I'd like to emphasize that the issue is relevant in the short term (1mo) too because it'll be harder to delete the adapter after `24.05`

SomeoneSerge commented on pkgs/stdenv/adapters.nix:
```diff
@@ -237,14 +237,13 @@ rec {
-        cc = targetStdenv.cc.cc;
-
+      /* NOTE(@connorbaker):
+       * This assumes targetStdenv.cc is a wrapped stdenv. */
```
```suggestion
       * This assumes targetStdenv.cc is a cc-wrapper. */
```

ConnorBaker replied:
Fixed in latest

ConnorBaker commented:
@SomeoneSerge I added your comment fix and included a reference to #283517. Can you take another look? Pending your approval and CI passing, I'd like to merge this.
